In [1]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris
# Load dataset Iris dari scikit-learn
iris = load_iris()

### Penjelasan:
In 1 bertujuan untuk mengimpor library utama yang dibutuhkan dalam analisis data, yaitu NumPy untuk komputasi numerik, Pandas untuk manipulasi struktur data, serta fungsi load_iris dari scikit-learn untuk memuat dataset standar Iris ke dalam memori.

In [2]:
# Konversi data menjadi DataFrame Pandas
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df["Species"] = iris.target_names[iris.target]
# Definisikan X (fitur) dan y (target/label)
X = df.drop(columns=["Species"])
y = df["Species"]
print(X)

     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]


### Penjelasan:
pada In 2 Data mentah dikonversi menjadi DataFrame. Lalu dilakukan pemisahan variabel: yaitu Variabel Bebas(X): yang terdiri dari 4 fitur kontinu, yaitu sepal length, sepal width, petal length, dan petal width, lalu Variabel Terikat(y): yg berupa label kategori spesies bunga Iris (setosa, versicolor, dan virginica).  

In [32]:
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

### Penjelasan:
y digunakan untuk mengevaluasi tipe dan nilai dari variabel target (y) yang masih dalam format kategorikal kualitatif (string/teks).  

In [33]:
from sklearn.preprocessing import LabelEncoder
lab = LabelEncoder()
y = lab.fit_transform(y)
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

### Penjelasan: 
In 4 ini Menggunakan teknik Label Encoding untuk mentransformasikan data kategorikal kualitatif pada variabel target menjadi representasi nilai numerik kuisensial (diskrit). Transformasi ini bertujuan agar variabel target dapat diproses secara matematis oleh algoritma pembelajaran mesin:   
1. Class setosa ->0
2. Class versicolor ->1
3. Class virginica ->2

In [34]:
from sklearn.model_selection import train_test_split
X_train, X_test = train_test_split(X, test_size=0.2, random_state=0)
y_train, y_test = train_test_split(y, test_size=0.2, random_state=0)

### Penjelasan:
 Pada in 5 dilakukan pemisahan dataset menjadi dua bagian independen menggunakan metode cross-validation split:   
 1. Data Latih (Training Set): Sebesar 80% dari total populasi data, digunakan untuk melatih model klasifikasi.   
 2. Data Uji (Testing Set): Sebesar 20% dari total populasi data, digunakan untuk mengevaluasi kinerja generalisasi model. 
 3. Parameter random_state=0 dipasang untuk menjamin keterulangan (reproducibility) hasil pembagian data

In [35]:
from sklearn.naive_bayes import GaussianNB
model = GaussianNB()
NB_model = model.fit(X_train, y_train)

### Penjelasan: 
In 6 Menginisialisasi algoritma Gaussian Naïve Bayes (GNB). Algoritma ini dipilih karena seluruh fitur dalam variabel bebas (X) merupakan data berdistribusi kontinu. Proses .fit() melakukan estimasi parameter statistik (mean $\mu$ dan standar deviasi $\sigma$) pada tiap fitur berdasar masing-masing kelas target pada training set.

In [36]:
result = NB_model.predict(X_test)
result

array([2, 1, 0, 2, 0, 2, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 1, 1, 0])

### Penjelasan: 
Mengaplikasikan model yang telah dilatih (NB_model) untuk melakukan inferensi (predict) terhadap variabel bebas pada testing set (X_{test}). Luaran yang dihasilkan berupa vektor prediksi kelas numerik (y_{pred}).

In [37]:
from sklearn.metrics import accuracy_score
print('Accuracy of data testing: ', accuracy_score(y_test, result)*100)

Accuracy of data testing:  96.66666666666667


### Penjelasan:
Menghitung matrik evaluasi kinerja berupa Akurasi Klasifikasi (Classification Accuracy), yaitu rasio antara jumlah prediksi benar terhadap total observasi data uji. Model menghasilkan nilai akurasi sebesar 96.67%, yang menunjukkan tingkat presisi inferensi tergolong sangat tinggi pada data uji.

In [38]:
#prediksi inputan data baru
#input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
input_data = [[5.1, 3.5, 1.4, 0.2]]
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)


Hasil prediksi untuk input baru adalah: [0]


c:\Users\User Acer\anaconda3\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


### Penjelasan:
Melakukan pengujian inferensi pada sampel data baru tak terlabel [5.1, 3.5, 1.4, 0.2]. Fungsi inverse_transform mengembalikan kode numerik hasil prediksi ke label nama spesies aslinya (['setosa']). Namun, eksekusi ini memicu UserWarning karena data masukan dioperasikan dalam struktur array 2D tanpa menyertakan metadata nama kolom/fitur (feature names).

In [39]:
input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: [0]


### Penjelasan: 
Melakukan perbaikan (refactoring) pada format masukan data baru dengan mengonversinya menjadi objek pd.DataFrame yang menyertakan header nama fitur iris.feature_names secara eksplisit. Langkah ini menyesuaikan struktur data sampel dengan struktur data saat pelatihan model, sehingga menghilangkan isu UserWarning dan memastikan konsistensi inferensi model.

In [40]:
# Tugas 2: 
input_data = pd.DataFrame(np.array([[7.3, 3.8, 4.6, 2.3]]), columns=iris.feature_names)
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)

print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: [2]


### Analisis:
Pertama, nilai ukuran bunga dibungkus ke dalam DataFrame Pandas lengkap dengan nama kolomnya agar strukturnya konsisten dengan data pelatihan dan mencegah timbulnya peringatan (warning). 

Selanjutnya, metode NB_model.predict digunakan untuk menghitung probabilitas dan menentukan kelas numerik dari data masukan tersebut. 

Fungsi lab.inverse_transform kemudian menerjemahkan kembali hasil kode numerik tersebut menjadi nama spesies aslinya.   

Dari proses kalkulasi ini, model menghasilkan output *virginica*, yang berarti data baru tersebut terklasifikasi sebagai spesies Iris virginica.